src/java.base/unix/classes/sun/nio/ch/FileDispatcherImpl.java

```diff
@@ -203,8 +206,10 @@ static void afterRestore() throws IOException {
     // Shared with SocketDispatcher and DatagramDispatcher but
     // NOT used by FileDispatcherImpl
     static void closeAndMark(FileDescriptor fd) throws IOException {
-        fdAccess.markClosed(fd);
-        close0(fd);
+        // Originally this used fdAccess.markClosed() and close0() but leaving
+        // the FD value set breaks JDKSocketResource (we don't want the extra
+        // test if the FD resource has been marked).
+        fdAccess.closeNoCleanup(fd);
```

Comment:
Do you mean this thread? https://mail.openjdk.org/pipermail/nio-dev/2023-January/013075.html

Well this was meant to change the behaviour as little as possible. All the ways the FD can be closed looks pretty spaghetti to me; I can give the regular close a go and see what breaks...

Reply:
Oh, sorry, I did not mean to make it more dangerous, just in the previous version it was rather inconsistent (the method is called closeAndMark, but apparently no marking was going to happen).

Thank you for the link. Does that mean that NIO does not reset FD just because that is not necessary?

And could you elaborate on why we don't want the extra test for the mark? Since we need more accurate tracking of the FD state, including close() from NIO, but we don't want to change the existing FD behavior, adding the check to JDKSocketResource would be pretty straightforward.

Reply:
> Thank you for the link. Does that mean that NIO does not reset FD just because that is not necessary?

I think so. It looks like a bit of stretch to me (writing a non-volatile field is not that costly) but I would think that someone did a benchmark and observed a difference.

> And could you elaborate on why we don't want the extra test for the mark?

To put it a bit into perspective, I have used the marking when only FileDescriptor's JDKResource was involved - one extra mark that would be set and read only by the resource seemed simple enough to do the job without changing the other code paths. But if the JDKSocketResource needs to break the encapsulation and check if (an independent) FD's resource has a mark set in order to treat it differently, the complexity builds up and isn't really worth of skipping the single field write.